### DISCUSSION TITLE
[Question][Hw3] Arguments of block relations and free_vars

### DISCUSSION THREAD

**Discussion by pingpingy1** [prosyslab-classroom/cs424-program-reasoning#327] [Q&A]
How are the arguemtns of the relations of each block specified?

I tried the following code to create the relation applications for each block:
```
let block_entry_relation bbenv bb =
  let { Verifier.relation = bb_relation; Verifier.free_vars = free_vars } = Verifier.BBMap.find bb bbenv in
  free_vars |> List.map Verifier.z3expr_of |> Z3.FuncDecl.apply bb_relation
```
but it resulted in an error stating that the sort of argument was not matched.
The error code is as follows:
```
Fatal error: exception Z3.Error("Sort mismatch at argument #2 for function (declare-fun => (Bool Bool) Bool) supplied sort is Int")
```
I wanted to ask if it indeed is the case that free_vars is the arguments needed for the relation created by the block.

Thank you.

**Comment by Hhro** [comment]
Yes, `freevars` are required to make a correct application of relation.
As you can see in `Verifier.env_of_function`:
```ocaml
let env_of_function cfg bbenv f = 
...
      let free_vars =
        VariableSet.diff reachable_vars bound_vars |> VariableSet.union phi_vars
      in
      let int_sort = Z3.Arithmetic.Integer.mk_sort z3ctx in
      let bool_sort = Z3.Boolean.mk_sort z3ctx in
      let sort_vec, free_vars =
        VariableSet.fold
          (fun var (sort_vec, free_vars) ->
            let lltype = Llvm.type_of var in
            match Llvm.classify_type lltype with
            | Llvm.TypeKind.Integer -> (
                match Llvm.integer_bitwidth lltype with
                | 32 -> (int_sort :: sort_vec, var :: free_vars)
                | 1 -> (bool_sort :: sort_vec, var :: free_vars)
                | _ -> failwith "Invalid type")
            | _ -> (sort_vec, free_vars))
          free_vars ([], [])
      in
      let relation =
        Z3.FuncDecl.mk_func_decl_s z3ctx
          (Llvm.value_name (Llvm.value_of_block bb))
          sort_vec bool_sort
      in
```
A `relation` is a function declaration defined as a sort vector sequentially mapped from `freevars` and bool sort. 

So, in fact, I don't see why your code raises an error.
If you insert the below code between the `let` expression and `free_var |> ... |> ... bb_relation` in your function, it will help your debugging.
```ocaml
List.iter (fun v -> v |> Z3.Expr.to_string |> print_endline) vars;
relation |> Z3.FuncDecl.to_string |> print_endline;
```

**Comment by pingpingy1** [comment]
Upon further inspection, it seems that this had little to do with the block reachability relation;
I messed up on a declaration of some other `implies` declaration.
Thank you very much for helping me out.

(Notice that in the error message, it says `(declare-fun => (Bool Bool) Bool)` instead of the name of some block.)
